---
title: "Introduction"
start: false
teaching: 10
exercises: 0
questions:
- "What type of data do we work on?"
- "How does the genome of *Zymoseptoria tritici* look like?"
objectives:
- "Describe the key characteristics of the genome of *Zymoseptoria tritici*"
keypoints:
- "*Z. tritici* is an important plant pathogen"
- "The genome of *Z. tritici* is ~40 Mb in size and has core chromosomes and dispensable chromosomes"
---

# Background of the data

In these exercises, we will analyse whole-genome sequences of a group of fungi from the order of Mycosphereallaceae. We will focus initially on *Zymoseptoria tritici*, also known as *Mycosphaerella graminicola*, the causal agent of Septoria leaf blotch on wheat. It is one of the most important diseases on wheat in Europe.

The genome of the *Z. tritici* reference genome of strain IPO323 was sequenced in 2011 [publications](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1002070). This genome was sequenced and then assembled using a combination of short-read sequencing data together with a genetic map. 

> ## Exercise
> 
> Take a brief look at the *Z. tritici* reference genome paper, especially table 1 and table 2. How many core chromosomes does the reference genome have, and how many dispensable (accessory) chromosomes? What is the main difference between core and dispensable chromosomes?
>
>> ## Solution
>> 
>> There are 13 core chromosomes and eight dispensable chromosomes. Dispensable chromosomes are shorter, have less genes and more repeats (transposons). 
>> 
> {: .solution}
{: .challenge}

In the last decade, many more *Z. tritici* isolates have been sequenced using a variety of sequencing methodologies (short-read and long-read sequencing technologies). An overview over some of the genomes we will be using can be found in the *Z. tritici* pan-genome [paper](https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-020-0744-3). 

# Outline of the coming days

In the first part of the practical, we will examine the quality of different genome assemblies of *Z. tritici*. We will specifically focus on differences between genome assemblies generated from sequencing data from 2nd or 3rd generation sequencing machines. In the afternoon, we will start comparing genome assemblies to identify large-scale structural differences between *Z. tritici* strains.

You will then focus on gene annotation and gene expression analyses in eukaryotes (Robin Ohm). After Robin's block, we will meet again to compare pan-genomes.

{% include links.md %}
